extensions/ql-vscode/src/view/variant-analysis/VariantAnalysisSkippedRepositoriesTab.tsx

```diff
@@ -0,0 +1,71 @@
+import * as React from 'react';
+import styled from 'styled-components';
+import { VariantAnalysisSkippedRepositoryGroup } from '../../remote-queries/shared/variant-analysis';
+import { Alert } from '../common';
+import { VariantAnalysisSkippedRepositoryRow } from './VariantAnalysisSkippedRepositoryRow';
+
+export type SkippedRepositoriesReason = 'no_access' | 'no_database';
+
+export type VariantAnalysisSkippedRepositoriesTabProps = {
+  reason: SkippedRepositoriesReason,
+  skippedRepositoryGroup: VariantAnalysisSkippedRepositoryGroup,
+};
+
+function getSkipReasonAlertTitle(reason: SkippedRepositoriesReason): string {
+  switch (reason) {
+    case 'no_access':
+      return 'No access';
+    case 'no_database':
+      return 'No database';
+  }
+}
+
+function getSkipReasonAlertMessage(
+  reason: SkippedRepositoriesReason,
+  repos: VariantAnalysisSkippedRepositoryGroup
+): string {
+  const repositoriesOmittedText = repos.repositoryCount > repos.repositories.length
+    ? ` (Only the first ${repos.repositories.length} ${repos.repositories.length > 1 ? 'repositories are' : 'repository is'} shown.)`
+    : '';
+  switch (reason) {
+    case 'no_access':
+      return `The following repositories could not be scanned because you do not have read access.${repositoriesOmittedText}`;
+    case 'no_database':
+      return `The following repositories could not be scanned because they do not have an available CodeQL database.${repositoriesOmittedText}`;
+  }
+}
+
+function getSkipReasonAlert(
+  reason: SkippedRepositoriesReason,
+  repos: VariantAnalysisSkippedRepositoryGroup
+) {
+  return (
+    <Alert
+      key='alert'
+      type='warning'
+      title={getSkipReasonAlertTitle(reason)}
+      message={getSkipReasonAlertMessage(reason, repos)}
+    />
+  );
+}
+
+const Container = styled.div`
+  display: flex;
+  flex-direction: column;
+  gap: 0.5em;
+  width: 100%;
+`;
+
+export const VariantAnalysisSkippedRepositoriesTab = ({
+  reason,
+  skippedRepositoryGroup,
+}: VariantAnalysisSkippedRepositoriesTabProps) => {
+  return (
+    <Container>
+      {getSkipReasonAlert(reason, skippedRepositoryGroup)}
+      {skippedRepositoryGroup.repositories.map((repo) =>
+        <VariantAnalysisSkippedRepositoryRow key={`repo/${repo.fullName}`} repository={repo} />
```

Comment:
The `key` only needs to be unique within its parent, so just using `repo.fullName` would be enough:

```suggestion
        <VariantAnalysisSkippedRepositoryRow key={repo.fullName} repository={repo} />
```

Reply:
My reasoning here was only to make super sure it doesn't conflict with the key for the alert, which is `alert`. Should be fine since a repo name contains a slash, but I didn't want to rely on that or risk conflicts if another component is added to this container. So unless there's more of a benefit to simplifying it I'd prefer to keep it as it is.

Reply:
I'm happy to leave it as-is, but I don't think you need the `key` on the alert at all since it's not part of the list. React only uses the keys for change detection in a `map` call/list. Even when the component is a sibling of a `map` call, it shouldn't be necessary to add the `key`.